### PR TITLE
Qute: if section - adjust the evaluation rules for equality operators

### LINKED ISSUE #44610: Qute: Unexpected primitive equality
### Describe the bug

This Qute template is rendered as `0 is not 0`:
```smarty
{@long number=0L}
{number} is {#if number == 0}0{#else}not 0{/if}
```
The same happens when comparing a `long` returned from a method with an `int` in a template.
I didn't check for other types.

### Expected behavior

Primitive numbers should be compared by their value, not their type.

### Actual behavior

`==` and `!=` fail with primitive numbers of different data types.

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

3.15.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by mkouba**
- fixes #44610

### CODE REVIEW

**Comment by mkouba** [comment]
@neon-dev Hopefully this makes a little bit more sense.

File: independent-projects/qute/core/src/main/java/io/quarkus/qute/IfSectionHelper.java [hunk 1]
```diff
@@ -553,25 +568,22 @@ static Operator from(String value) {
...
-            } else {
-                throw new TemplateException("Not a valid number: " + value);
-            }
-            return decimal;
+            if (value instanceof BigDecimal decimal) {
+                return decimal;
+            } else if (value instanceof BigInteger bigInteger) {
+                return new BigDecimal(bigInteger);
+            } else if (value instanceof Integer integer) {
+                return new BigDecimal(integer);
```
Comment (gastaldi) [inline independent-projects/qute/core/src/main/java/io/quarkus/qute/IfSectionHelper.java:576]: Use `BigDecimal.valueOf` to take advantage of the internal caching?

Comment (mkouba) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/IfSectionHelper.java:576]: Good point.

File: independent-projects/qute/core/src/main/java/io/quarkus/qute/IfSectionHelper.java [hunk 2]
```diff
@@ -492,6 +492,21 @@ boolean evaluate(Object op1, Object op2) {
...
+            if (op1 != null && op2 != null && (op1 instanceof Number || op2 instanceof Number)) {
+                if (op1.getClass().equals(op2.getClass())) {
+                    // Both operands are of the same Number type
+                    return op1.equals(op2);
+                }
+                // Both operands are not null and at least one of them is a number
+                return getDecimal(op1).compareTo(getDecimal(op2)) == 0;
+            }
+            return Objects.equals(op1, op2);
+        }
```
[the remark this answers is not in the document]
Comment (mkouba) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/IfSectionHelper.java:508]: Hm yes, well, your proposal is not functionally equivalent. The current version allows for comparison of e.g. `String` and `Integer`. Also `op1 == op2` should be, in theory, more effective than `equals()`. In other words, we only want to call `equals` unless the objects are identical or number comparison is possible. Does that make sense?

[the remark this answers is not in the document]
Comment (neon-dev) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/IfSectionHelper.java:508]: > The current version allows for comparison of e.g. String and Integer.

Oh yeah, I completely missed that. So `{#if 0 == "0"}` should be `true` if I understand correctly?

> we only want to call equals unless the objects are identical or number comparison is possible

`Objects.equals` is implemented as `(a == b) || (a != null && a.equals(b));`, so we would only see unnecessary `.equals` calls for two objects of different Number types.

[the remark this answers is not in the document]
Comment (mkouba) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/IfSectionHelper.java:508]: > > The current version allows for comparison of e.g. String and Integer.
> 
> Oh yeah, I completely missed that. So `{#if 0 == "0"}` should be `true` if I understand correctly?

Yes, I think so. We do the same for comparison operators such as `>`; e.g. `{#if 10 > "1"}` should be also true. 

> > we only want to call equals unless the objects are identical or number comparison is possible
> 
> `Objects.equals` is implemented as `(a == b) || (a != null && a.equals(b));`, so we would only see unnecessary `.equals` calls for two objects of different Number types.

Yes.

**Review by gastaldi — APPROVED** [review]

File: independent-projects/qute/core/src/main/java/io/quarkus/qute/IfSectionHelper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkouba) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/IfSectionHelper.java:508]: Yes, a little bit better in terms of readability but a little worse in terms of performance. 

Let's prioritize readability ;-).

[the remark this answers is not in the document]
Comment (mkouba) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/IfSectionHelper.java]: Done.

**Comment by mkouba** [comment]
It seems that `io.quarkus.cli.CliProjectJBangTest` failed with `Error downloading JBang`. Not sure what's going on but it's definitely not related to this PR.

**Pull request merged**